contracts/cw-names-registry/src/contract.rs

```diff
@@ -0,0 +1,256 @@
+
+use crate::error::ContractError;
+use crate::msg::{
+    ExecuteMsg, InstantiateMsg, LookUpDaoResponse, LookUpNameResponse, QueryMsg, ReceiveMsg,
+};
+use crate::state::{Config, CONFIG, DAO_TO_NAME, NAME_TO_DAO, RESERVED_NAMES};
+
+const CONTRACT_NAME: &str = "crates.io:cw-name-registry";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let payment_token_address = deps.api.addr_validate(&msg.payment_token_address)?;
+    let admin = deps.api.addr_validate(&msg.admin)?;
+
+    if msg.payment_amount.is_zero() {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let config = Config {
+        admin,
+        payment_token_address,
+        payment_amount: msg.payment_amount,
+    };
+
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::new().add_attribute("action", "instantiate"))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(wrapped) => execute_receive(deps, env, info, wrapped),
+        ExecuteMsg::UpdateConfig {
+            new_payment_token_address,
+            new_admin,
+            new_payment_amount,
+        } => execute_update_config(
+            deps,
+            env,
+            info,
+            new_payment_token_address,
+            new_admin,
+            new_payment_amount,
+        ),
+        ExecuteMsg::Revoke { name } => execute_revoke(deps, env, info, name),
+        ExecuteMsg::Reserve { name } => execute_reserve(deps, env, info, name),
+        ExecuteMsg::TransferReservation { name, dao } => {
+            execute_transfer_reservation(deps, env, info, name, dao)
+        }
+    }
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // We only take payments from our specified token
+    if info.sender != config.payment_token_address {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // TODO: Is there a way we can verify this is a DAO
+    let sender = wrapped.sender;
+    let amount = wrapped.amount;
+    let msg: ReceiveMsg = from_binary(&wrapped.msg)?;
+
+    match msg {
+        ReceiveMsg::Register { name } => register_name(deps, env, sender, amount, name),
+    }
+}
+
+pub fn register_name(
+    deps: DepsMut,
+    _env: Env,
+    sender: String,
+    amount: Uint128,
+    name: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if amount < config.payment_amount {
```

Comment:
See note above - `if amount < config.payment_required_to_register_name` is a bit more self-explanatory.